### ISSUE TITLE
In 2024, Android/ExoPlayer still does not support transparent (alpha) video? Even though Google created VP8/VP9 which support transparency? Any help or solution?

### ISSUE THREAD

**Issue by jonmdev** [androidx/media#1388]
### Use case description
Android is to my knowledge the only major operating system which in 2024 still does not support video playback of transparent (alpha) video. ExoPlayer is the main recommended video player of Android. 

People have been requesting this feature in the following thread since 2020:

https://github.com/google/ExoPlayer/issues/7789

We have heard nothing formally from Google or Android about why this is not supported or if there will ever be support for it or when. 

Google is the primary developer for VP8 and VP9 which it seeks to make standard codecs for broad usage. Both support transparent backgrounds. But yet Android cannot in any way play back transparent video using these codecs.

This does not seem logical. It would be hoped that someone with power at Google could assign a few engineers to fix ExoPlayer so we can utilize the very transparent video codecs Google has developed for us.

### Proposed solution

Please allow the TextureView mode of ExoPlayer to playback transparent video (alpha video support, eg. VP8/VP9) such that a transparent region of the video will show the view underneath as would be expected.

### Alternatives considered

The only existing workaround for the lack of transparency support was proposed here:
https://medium.com/go-electra/unlock-transparency-in-videos-on-android-5dc43776cc72

This individual:
- Created custom video with a side by side view (rgb channel on one side, and alpha channel on other side in gray levels)
- Played this video using ExoPlayer
- Outputted ExoPlayer into a SurfaceTexture
- Used an OpenGL shader to merge the data of the SurfaceTexture to output rgb+a pixels
- Rendered the shader output on a TextureView

The amount of technical expertise and custom coding required to do all this is absurd considering again that Google's flagship video codecs VP8/VP9 already support transparency. No other operating system requires such pain to render a transparent video.

It is Google's intent that we also follow the same approach of creating SurfaceTextures and custom OpenGL shaders to merge the SurfaceTextures and render to a TextureView?

Or is it logical that ExoPlayer should be fixed and upgraded to support the basic VP8/VP9 feature set?

**Comment by marcbaechinger** [comment]
Thanks for your report!

> Android cannot in any way natively play back transparent video using these codecs

You are correct that playing media efficiently with ExoPlayer requires codec support from the operating system it runs on. 

I'm not sure if this issue tracker is the right place to track framework codec feature development, but I keep this issue open for visibility. We as the Media3 team can't really help you with this expect by raising awareness for the issue. I actually think this is a known issue and I'm not involved into codec development and the reasoning behind this or that roadmap. However, I'll make sure the codec team is aware of this issue even though I think this is already the case.

**Comment by marcbaechinger** [comment]
Internal bug ref: 190802665

**Comment by willsmanley** [comment]
if it helps for context, this is also blocking cross-platform adoption for transparent video in frameworks like flutter, where the video players rely on exoplayer for android builds. 

@jonmdev one workaround you might want to try in the meantime is using a webview since the browsers do support alpha channels

**Comment by jonmdev** [comment]
> one workaround you might want to try in the meantime is using a webview since the browsers do support alpha channels

Yes that is exactly what I did, though I think it is insanely foolish that this is required. If the web view can render transparent video, why not Android directly? Doesn't add up.

It is not like Google is some poor volunteer run company with no resources.

**Comment by andrewlewis** [comment]
I've asked for an update from the framework team on the internal bug. I'm afraid we don't have an efficient way to support this in media3 until the underlying limitations in the framework are removed.

**Comment by willsmanley** [comment]
one other solution for anybody looking - if your transparency is always in the same place, you can use a mask to clip the transparent parts. clearly this wouldnt work if your transparency moves throughout the duration of the video, but this ended up being a smoother solution for me than webview (i had an opaque circle with transparent background). hopefully this helps anyone else trying to solve this in the meantime. took me >1 month to think of this retrospectively obvious solution.

**Comment by Psijic** [comment]
Hello, I see the issue is still opened. Can I play a transparent video in exoplay in 2025 without ad-hocks? I need to set the alpha channel color like Color.Black and voila - it's played with no background. Is it possible?

**Comment by ShaifMini** [comment]
+1

**Comment by mhaslinsky** [comment]
+1

**Comment by ivanksk-creditgenie** [comment]
+1

**Comment by israellewis-lab** [comment]
+1

**Comment by ozgun-ozkan** [comment]
+1

**Comment by SmartPlayerDS** [comment]
+1

**Comment by ShaifMini** [comment]
+1

**Comment by manas-dadheech-pop** [comment]
+1 videos that play fine on chrome do not play properly with exo player if they have alpha channel. same is observed on google drive as well.

**Comment by nioto22** [comment]
+1

**Comment by 6xingyv** [comment]
Find a branch [here](https://github.com/zxzx74147/ExoPlayer), i don't know if they can start a PR...

**Comment by KamenGoranchev** [comment]
+1

**Comment by Sk1droWWW** [comment]
+1

**Comment by alinimbuzzer** [comment]
It is 2026 still there is no alpha channel support, because of this single issue we have to move again  from kotlin jetpack compose.

**Comment by 6xingyv** [comment]
> It is 2026 still there is no alpha channel support, because of this single issue we have to move again from kotlin jetpack compose.

For now, my workaround might be useful.
https://github.com/6xingyv/compound/tree/master/app/src/main/cpp
https://github.com/6xingyv/compound/blob/master/app/src/main/java/com/mocharealm/compound/ui/composable/base/VpxVideoPlayer.kt

**Comment by agonist** [comment]
Looks like Skynet will take over the world before we're able to play transparent videos on Android

**Comment by Sk1droWWW** [comment]
Judging by the looks of it, the following scenario will play out before this issue even makes it onto some Google engineer's to-do list:

Sooner or later, someone will join Google, fix these issues in a week, and then leave Google to continue working on a project that requires transparent video

**Comment by andrewlewis** [comment]
Sorry for the lack of updates. About Android OS support for this: the current update we have from the codec framework team is "the blockers are basically having transparent profiles listed in the codec APIs, so they can be queried for support, and then adding compliance testing to verify correct decoding of alpha-channel videos. Possibly adding more alpha-channel image formats as well, not sure if RGBA would cover all the use cases or if YUV+Alpha would also be desirable", and from the graphics team "We've [...] enhanced SurfaceView to support alpha blending (via SurfaceView#setAlpha). We did that in Android 14. I'm not aware of any inherent limitation for per-pixel alpha channels". I'm regularly bringing up the requests here and from other channels but I don't think the codec framework changes are on the roadmap for the next release. I'm hopeful that now we have LLMs that may make it easier to prioritize and I will ping them again to ask.

On the Jetpack side, in the 2026 roadmap we plan to add infrastructure for handling multiple video tracks where samples are timestamp-aligned and rendered together. This can be used to implement alpha video support, as it will make it easier to hande content where color and translucency are in two timestamp-aligned video tracks. Then the next step would be to 'join' these at render time. With this approach I don't expect there to be requirements on the codec framework (though it may depend on exactly how the video with alpha is authored).
